### ISSUE TITLE
sync: unlocking an unlocked Mutex panics (should throw)

### ISSUE THREAD

**Issue by rhysh** [golang/go#23039]
### What version of Go are you using (`go version`)?

`go version go1.9.2 darwin/amd64`
`go version devel +28736053ad Wed Dec 6 15:50:04 2017 +0000 darwin/amd64`

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

`darwin/amd64`, and the Go playground

### What did you do?

```
defer func() { recover() }()
new(sync.Mutex).Unlock()
```

`sync.Mutex`: https://play.golang.org/p/7MFt4Gn037
`sync.RWMutex`: https://play.golang.org/p/YWnu02fr4X

### What did you expect to see?

I expected unlocking an unlocked sync.Mutex to lead to an unrecoverable fatal error, as it did in Go 1.8 and as unlocking an unlocked sync.RWMutex does today.

### What did you see instead?

A couple of calls to `throw` were converted to `panic` calls as part of https://github.com/golang/go/commit/0556e26273f704db73df9e7c4c3d2e8434dec7be, apparently unintentionally undoing a portion of https://github.com/golang/go/commit/40d81cf061d8a2a277d70446f582a984c1701ff3.

```
$ cat /tmp/bug.go 
package main

import "sync"

func main() {
        defer func() { recover() }()
        new(sync.Mutex).Unlock()
}
$ go1.7 run /tmp/bug.go
$ go1.8 run /tmp/bug.go
fatal error: sync: unlock of unlocked mutex

goroutine 1 [running]:
runtime.throw(0x10696e9, 0x1e)
        /usr/local/go/src/runtime/panic.go:596 +0x95 fp=0xc42003ff20 sp=0xc42003ff00
sync.throw(0x10696e9, 0x1e)
        /usr/local/go/src/runtime/panic.go:585 +0x35 fp=0xc42003ff40 sp=0xc42003ff20
sync.(*Mutex).Unlock(0xc42000e0d0)
        /usr/local/go/src/sync/mutex.go:113 +0xa4 fp=0xc42003ff68 sp=0xc42003ff40
main.main()
        /tmp/bug.go:7 +0x57 fp=0xc42003ff88 sp=0xc42003ff68
runtime.main()
        /usr/local/go/src/runtime/proc.go:185 +0x20a fp=0xc42003ffe0 sp=0xc42003ff88
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc42003ffe8 sp=0xc42003ffe0
exit status 2
$ go1.9 run /tmp/bug.go
$ go-tip run /tmp/bug.go
```

**Comment by bradfitz** [comment]
Probably too late for Go 1.10, especially since we shipped Go 1.9 with this, but I'm not overly concerned if it goes into Go 1.10 either.

/cc @dvyukov @ianlancetaylor

**Comment by bradfitz** [comment]
Also @rsc.
